lib/core/src/Cardano/Wallet/Primitive/AddressDiscovery/Script.hs

```diff
@@ -0,0 +1,176 @@
+{-# LANGUAGE LambdaCase #-}
+{-# LANGUAGE MultiParamTypeClasses #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE TypeApplications #-}
+{-# LANGUAGE TypeFamilies #-}
+{-# LANGUAGE UndecidableInstances #-}
+
+{-# OPTIONS_GHC -fno-warn-orphans #-}
+
+-- |
+-- Copyright: © 2018-2021 IOHK
+-- License: Apache-2.0
+--
+-- An implementation of shared script state using
+-- scheme specified in CIP-XXX Multi-signature Wallets.
+
+module Cardano.Wallet.Primitive.AddressDiscovery.Script
+    (
+      keyHashFromAccXPubIx
+    , constructAddressFromIx
+    , toNetworkTag
+    , replaceCosignersWithVerKeys
+    , liftPaymentAddress
+    , liftDelegationAddress
+    ) where
+
+import Prelude
+
+import Cardano.Address.Script
+    ( Cosigner (..)
+    , KeyHash (..)
+    , Script (..)
+    , ScriptHash (..)
+    , ScriptTemplate (..)
+    , toScriptHash
+    )
+import Cardano.Address.Style.Shelley
+    ( Credential (..)
+    , delegationAddress
+    , deriveMultisigForDelegationPublicKey
+    , deriveMultisigForPaymentPublicKey
+    , hashKey
+    , liftXPub
+    , mkNetworkDiscriminant
+    , paymentAddress
+    )
+import Cardano.Crypto.Wallet
+    ( XPub )
+import Cardano.Wallet.Primitive.AddressDerivation
+    ( Depth (..)
+    , DerivationType (..)
+    , Index (..)
+    , KeyFingerprint (..)
+    , NetworkDiscriminant (..)
+    , SoftDerivation
+    , WalletKey (..)
+    , deriveVerificationKey
+    , hashVerificationKey
+    )
+import Cardano.Wallet.Primitive.Types.Address
+    ( Address (..) )
+import Data.Either.Combinators
+    ( fromRight', rightToMaybe )
+import Data.Maybe
+    ( fromMaybe )
+import Data.Type.Equality
+    ( (:~:) (..), testEquality )
+import Type.Reflection
+    ( Typeable, typeRep )
+
+import qualified Cardano.Address as CA
+import qualified Cardano.Address.Style.Shelley as CA
+import qualified Data.Map.Strict as Map
+
+keyHashFromAccXPubIx
+    :: (SoftDerivation k, WalletKey k)
+    => k 'AccountK XPub
+    -> Index 'Soft 'ScriptK
+    -> KeyHash
+keyHashFromAccXPubIx accXPub ix =
+    hashVerificationKey $ deriveVerificationKey accXPub ix
+
+replaceCosignersWithVerKeys
+    :: CA.Role
+    -> ScriptTemplate
+    -> Index 'Soft 'ScriptK
+    -> Script KeyHash
+replaceCosignersWithVerKeys role (ScriptTemplate xpubs scriptTemplate) ix =
+    replaceCosigner scriptTemplate
+  where
+    replaceCosigner :: Script Cosigner -> Script KeyHash
+    replaceCosigner = \case
+        RequireSignatureOf c -> RequireSignatureOf $ toKeyHash c
+        RequireAllOf xs      -> RequireAllOf (map replaceCosigner xs)
+        RequireAnyOf xs      -> RequireAnyOf (map replaceCosigner xs)
+        RequireSomeOf m xs   -> RequireSomeOf m (map replaceCosigner xs)
+        ActiveFromSlot s     -> ActiveFromSlot s
+        ActiveUntilSlot s    -> ActiveUntilSlot s
+    toKeyHash :: Cosigner -> KeyHash
+    toKeyHash c =
+        let ix' = toEnum (fromEnum ix)
+            (Just accXPub) = liftXPub <$> Map.lookup c xpubs
```

Comment:
Is this pattern-match safe? If it is, the function should state why and possibly, make explicit any invariant that are expected of the ScriptTemplate.

Reply:
yes, it is safe as we can only do it when each cosigner has corresponding accXPub collected. nevertheless, introduces invariant in https://github.com/input-output-hk/cardano-wallet/pull/2508/commits/37e818cc20a5248c85f8645e712295541941e614